solidity/contracts/deposit/DepositUtils.sol

```diff
@@ -458,56 +464,70 @@ library DepositUtils {
     /// @param _redeemer    The deposit redeemer.
     /// @return             The amount in TBTC needed to redeem the deposit.
     function getRedemptionTbtcRequirement(DepositUtils.Deposit storage _d, address _redeemer) internal view returns(uint256) {
-        uint256 mainCharge = mainCharge(_d, _redeemer);
-        uint256 frtAdjust = frtAdjust(_d, _redeemer);
-        uint256 signerFeeAdjust = signerFeeAdjust(_d, _redeemer);
-        return mainCharge.add(frtAdjust).add(signerFeeAdjust);
-    }
+        bool tdtOwnerRedemption = depositOwner(_d) == _redeemer;
+        bool frtOwnerRedemption = feeRebateTokenHolder(_d) == _redeemer;
+        bool frtExists = feeRebateTokenHolder(_d) != address(0);
 
-    /// @notice             Get the base TBTC amount needed to redeem.
-    /// @param _redeemer    The deposit redeemer.
-    /// @return             The amount in TBTC.
-    function mainCharge(DepositUtils.Deposit storage _d, address _redeemer) internal view returns (uint256){
         bool inCourtesy = _d.inCourtesyCall();
-        bool ownerRedemption = depositOwner(_d) == _redeemer;
         bool atTerm = remainingTerm(_d) == 0;
+        bool redeemerPaysNoFee = inCourtesy || atTerm;
+
+        require(inCourtesy || tdtOwnerRedemption || atTerm, "Only TDT holder can redeem unless deposit is at-term or in COURTESY_CALL");
```

Comment:
Let's put the owner redemption check up front, and I'm interested here as well in a common structure to the others, which would be `redeemerHoldsTdt` or similar.